### ISSUE TITLE
[Victorialogs] panic: BUG: too big number of columns detected in the block

### ISSUE THREAD

**Issue by deemon87** [VictoriaMetrics/VictoriaMetrics#7568]
### Describe the bug

panic: BUG: too big number of columns detected in the block: 2177; the number of columns mustn't exceed 2000

### To Reproduce

I don't know how do it

### Version

```
victoria-logs-20241112-181012-tags-v1.0.0-victorialogs-0-g99607e2f3b
```

### Logs

```
Nov 18 12:05:01 victoria-logs-prod[209734]:         net/http/server.go:3360 +0x485
Nov 18 12:05:01 victoria-logs-prod[209734]: created by net/http.(*Server).Serve in goroutine 162
Nov 18 12:05:01 victoria-logs-prod[209734]:         net/http/server.go:2092 +0x5d0
Nov 18 12:05:01 victoria-logs-prod[209734]: net/http.(*conn).serve(0xc0000e2480, {0xc4eed0, 0xc00608a750})
Nov 18 12:05:01 victoria-logs-prod[209734]:         net/http/server.go:3210 +0x8e
Nov 18 12:05:01 victoria-logs-prod[209734]: net/http.serverHandler.ServeHTTP({0xc4c0c0?}, {0xc4df48?, 0xc098004540?}, 0x6?)
Nov 18 12:05:01 victoria-logs-prod[209734]:         net/http/server.go:2220 +0x29
Nov 18 12:05:01 victoria-logs-prod[209734]: net/http.HandlerFunc.ServeHTTP(0x40d8e5?, {0xc4df48?, 0xc098004540?}, 0xc098004501?)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/klauspost/compress@v1.17.10/gzhttp/compress.go:500 +0x5ef
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/klauspost/compress/gzhttp.NewWrapper.func1.1({0xc4df48, 0xc098004540}, 0xc098028140)
Nov 18 12:05:01 victoria-logs-prod[209734]:         net/http/server.go:2220 +0x29
Nov 18 12:05:01 victoria-logs-prod[209734]: net/http.HandlerFunc.ServeHTTP(0x117a8c0?, {0xc4e818?, 0xc00ced86e0?}, 0x4?)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:249 +0x31
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.gzipHandler.func1({0xc4e818?, 0xc00ced86e0?}, 0xc00ced86e0?)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:415 +0xf4f
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper(0xc005106400, {0xc4e818, 0xc00ced86e0}, 0xc098028140, 0xbb2b30)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/app/victoria-logs/main.go:89 +0x17d
Nov 18 12:05:01 victoria-logs-prod[209734]: main.requestHandler({0xc4e038, 0xc096a745d0}, 0xc098028140)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert/main.go:44 +0x196
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert.RequestHandler({0xc4e038, 0xc096a745d0}, 0xc098028140)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:107 +0x715
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert/elasticsearch.RequestHandler({0xc096a685ba, 0x6}, {0xc4e038, 0xc096a745d0}, 0xc098028140)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert/insertutils/common_params.go:222 +0x3a
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert/insertutils.(*logMessageProcessor).MustClose(0xc096a7e8c0)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert/insertutils/common_params.go:213 +0x52
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/app/vlinsert/insertutils.(*logMessageProcessor).flushLocked(0xc096a7e8c0)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vlstorage/main.go:133
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/app/vlstorage.MustAddRows(...)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/storage.go:517 +0x831
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*Storage).MustAddRows(0xc0000d2420, 0xc006b28800)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/partition.go:142 +0x310
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*partition).mustAddRows(0xc000866fc0, 0xc006b28800)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:675 +0x89
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustAddRows(0xc000624240, 0xc006b28800)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/inmemory_part.go:88 +0x1e5
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*inmemoryPart).mustInitFromRows(0xc02bdfe640, 0xc006b28800)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:338 +0xb9
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).MustWriteRows(0xc01f78e000, 0xc0abe1a030, {0xc0ad19d000, 0x24b, 0xa00}, {0xc0ad1ac000, 0x24b, 0x955})
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:227 +0x2d3
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*block).MustInitFromRows(0xc001086050, {0xc0ad19d000, 0x24b, 0xa00}, {0xc0ad1ac000, 0x24b, 0x955})
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:370 +0x92
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*block).sortColumnsByName(0xc001086050)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:143
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.Panicf(...)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:147
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevel(...)
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:155 +0x1a5
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevelSkipframes(0x1, {0xa62213, 0x5}, {0xa917b4?, 0xffffffffffffffff?}, {0xc0051b8f30?, 0xc09d2254a8?, 0xc0051b8fc8?})
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:315 +0xa74
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logMessage({0xa62213, 0x5}, {0xc096a5ab60, 0x65}, 0x4)
Nov 18 12:05:01 victoria-logs-prod[209734]:         runtime/panic.go:785 +0x132
Nov 18 12:05:01 victoria-logs-prod[209734]: panic({0x9ccee0?, 0xc098020430?})
Nov 18 12:05:01 victoria-logs-prod[209734]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:290 +0x5b
Nov 18 12:05:01 victoria-logs-prod[209734]: github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper.func1()
Nov 18 12:05:01 victoria-logs-prod[209734]: goroutine 198 [running]:
Nov 18 12:05:01 victoria-logs-prod[209734]: panic: BUG: too big number of columns detected in the block: 2177; the number of columns mustn't exceed 2000
Nov 18 12:05:01 victoria-logs-prod[209734]: 2024-11-18T12:05:01.970Z        panic        VictoriaMetrics/lib/logstorage/block.go:370        BUG: too big number of columns detected in the block: 2177; the number of columns mustn't exceed 2000
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.323Z        info        VictoriaMetrics/lib/httpserver/httpserver.go:133        pprof handlers are exposed at http://0.0.0.0:9122/debug/pprof/
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.322Z        info        VictoriaMetrics/lib/httpserver/httpserver.go:132        started server at http://0.0.0.0:9122/
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.304Z        info        VictoriaMetrics/app/victoria-logs/main.go:50        started VictoriaLogs in 0.106 seconds; see https://docs.victoriametrics.com/victorialogs/
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.304Z        info        VictoriaMetrics/app/vlstorage/main.go:70        successfully opened storage in 0.106 seconds; smallParts: 89; bigParts: 87; smallPartBlocks: 13853; bigPartBlocks: 752363; smallPartRows: 5740956; bigPartRows: 1288978660; smallPartSize: 566008108 bytes; bigPartSize: 104590672142 bytes
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.199Z        info        VictoriaMetrics/lib/memory/memory.go:42        limiting caches to 10307921510 bytes, leaving 6871947674 bytes to the OS according to -memory.allowedPercent=60
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/app/vlstorage/main.go:64        opening storage at -storageDataPath=/opt/victorialogs/data_dir
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/app/victoria-logs/main.go:42        starting VictoriaLogs at "[:9122]"...
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/lib/logger/flag.go:20          -storageDataPath="/opt/victorialogs/data_dir"
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/lib/logger/flag.go:20          -retentionPeriod="14d"
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/lib/logger/flag.go:20          -insert.maxLineSizeBytes="16MB"
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/lib/logger/flag.go:20          -httpListenAddr=":9122"
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/lib/logger/flag.go:20          -envflag.enable="true"
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/lib/logger/flag.go:13        command-line flags
Nov 18 12:05:00 victoria-logs-prod[209734]: 2024-11-18T12:05:00.197Z        info        VictoriaMetrics/lib/logger/flag.go:12        build version: victoria-logs-20241112-181012-tags-v1.0.0-victorialogs-0-g99607e2f3b
```

**Comment by jiekun** [comment]
Thanks! Let me try and see if I could reproduce it locally with your config example.

**Comment by jiekun** [comment]
Thanks! so in other words, there are really so many fields, and it is not related to the Elasticsearch sink, it's an issue with the source data?

**Comment by f41gh7** [comment]
Thanks for reporting, issue is reproduced and soon should be fixed.

**Comment by valyala** [comment]
The bug should be fixed by the commit 22e6385f56afcd518854ca9b97a5068736be4333 . This commit will be included in the next release. In the mean time it is possible to build VictoriaLogs from this commit according to [these instructions](https://docs.victoriametrics.com/victorialogs/quickstart/#building-from-source-code) and verifying whether the bug is fixed there.

Also, an additional commit - 4d39dfc0ee196cb4d6e3feac0d012ab141403261 - improves logging for this kind of error in the future - now the error message will contain the list of log field names, which exceeds the limit on the number of unique fields per stored block of data. The list of log field names can simplify detecting the root cause of too big number of unique log field names.

**Comment by valyala** [comment]
> Also, I think, that configuration param -insert.maxFieldsPerLine must be removed or reworked. If change this value greater than maxColumnsPerBlock (2000) and ingest log line that exceeds given limit. E.g. single log line with unique fields. VictoriaLogs could panic at ingestion.

This has been addressed in the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/6a7192156598862818e0c751269d6fe9cae1249c . Now VictoriaLogs skips log entries with too many log fields during data ingestion. It also emits a warning to logs with the list of fields in the skipped log entry, so the user could identify the source of logs and fix it ASAP.

This commit and the commit mentioned above are included in the [v1.1.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/) release. Closing the issue as fixed then.
